wazo_webhookd/plugins/subscription/bus.py

```diff
@@ -59,23 +58,46 @@ def _build_headers(subscription):
     def _register(self, subscription):
         uuid = subscription.uuid
         data = subscription_schema.dump(subscription)
-        extra_headers = self._build_headers(subscription)
 
-        fn, events = self._subscription_callbacks[uuid] = (
+        fn, events, extra_headers = self._subscription_callbacks[uuid] = (
             partial(self._callback, data),
             subscription.events,
+            self._build_headers(subscription),
         )
 
         for event in events:
             self._bus_consumer.subscribe(event, fn, headers=extra_headers)
 
     def _unregister(self, subscription):
         uuid = subscription.uuid
-        fn, events = self._subscription_callbacks.pop(uuid, (None, []))
+        fn, events, _ = self._subscription_callbacks.pop(uuid, (None, []))
 
         for event in events:
             self._bus_consumer.unsubscribe(event, fn)
 
+    def _update(self, subscription):
+        uuid = subscription.uuid
+        data = subscription_schema.dump(subscription)
+        headers = self._build_headers(subscription)
+
+        prev_fn, prev_events, prev_headers = self._subscription_callbacks.pop(uuid)
+        fn = partial(self._callback, data)
+
+        all_events = set(subscription.events) | set(prev_events)
+
+        for event in all_events:
+            if event not in subscription.events:  # removed events
+                self._bus_consumer.unsubscribe(event, prev_fn)
+            elif event not in prev_events:  # newly added event
+                self._bus_consumer.subscribe(event, fn, headers=headers)
+            elif headers != prev_headers:  # headers changed
+                self._bus_consumer.subscribe(event, fn, headers=headers)
+                self._bus_consumer.unsubscribe(event, prev_fn)
+            else:  # only update callback
+                self._bus_consumer.unsubscribe(event, prev_fn)
+                self._bus_consumer.subscribe(event, fn, headers=headers)
+        self._subscription_callbacks[uuid] = (fn, subscription.events, headers)
```

Comment:
I like your algorithm :)

Don't we need a lock around `self._subscription_callbacks` to avoid concurrent updates?

Reply:
You are right, although because of the implementation of the GIL, dict are implicitly thread-safe (in CPython), it is a better practice to make it explicit, thanks!